pkg/agent/qrm-plugins/memory/handlers/sockmem/sockmem_linux.go

```diff
@@ -0,0 +1,204 @@
+//go:build linux
+// +build linux
+
+/*
+Copyright 2022 The Katalyst Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package sockmem
+
+import (
+	"context"
+	"fmt"
+
+	"golang.org/x/sys/unix"
+
+	coreconfig "github.com/kubewharf/katalyst-core/pkg/config"
+	dynamicconfig "github.com/kubewharf/katalyst-core/pkg/config/agent/dynamic"
+	coreconsts "github.com/kubewharf/katalyst-core/pkg/consts"
+	"github.com/kubewharf/katalyst-core/pkg/metaserver"
+	"github.com/kubewharf/katalyst-core/pkg/metaserver/agent/metric/helper"
+	"github.com/kubewharf/katalyst-core/pkg/metrics"
+	"github.com/kubewharf/katalyst-core/pkg/util/cgroup/common"
+	cgroupcm "github.com/kubewharf/katalyst-core/pkg/util/cgroup/common"
+	cgroupmgr "github.com/kubewharf/katalyst-core/pkg/util/cgroup/manager"
+	"github.com/kubewharf/katalyst-core/pkg/util/general"
+	"github.com/kubewharf/katalyst-core/pkg/util/native"
+)
+
+type SockMemConfig struct {
+	globalTCPMemRatio int
+	cgroupTCPMemRatio int
+}
+
+var sockMemConfig = SockMemConfig{
+	globalTCPMemRatio: 20,  // default: 20% * {host total memory}
+	cgroupTCPMemRatio: 100, // default: 100% * {cgroup memory limit}
+}
+
+func updateGlobalTCPMemRatio(ratio int) {
+	if ratio < globalTCPMemRatioMin {
+		ratio = globalTCPMemRatioMin
+	} else if ratio > globalTCPMemRatioMax {
+		ratio = globalTCPMemRatioMax
+	}
+	sockMemConfig.globalTCPMemRatio = ratio
+}
+
+func updateCgroupTCPMemRatio(ratio int) {
+	if ratio < cgroupTCPMemRatioMin {
+		ratio = cgroupTCPMemRatioMin
+	} else if ratio > cgroupTCPMemRatioMax {
+		ratio = cgroupTCPMemRatioMax
+	}
+	sockMemConfig.cgroupTCPMemRatio = ratio
+}
+
+func setHostTCPMem(memTotal uint64) error {
+	tcpMemRatio := sockMemConfig.globalTCPMemRatio
+	tcpMem, err := getHostTCPMemFile(hostTCPMemFile)
+	if err != nil {
+		fmt.Println("Error:", err)
+		return err
+	}
+
+	pageSize := uint64(unix.Getpagesize())
+	newUpperLimit := memTotal / pageSize / 100 * uint64(tcpMemRatio)
+	if (newUpperLimit != tcpMem[2]) && (newUpperLimit > tcpMem[1]) {
+		general.Infof("write to host tcp_mem, ratio=%d, newLimit=%d, oldLimit=%d", tcpMemRatio, newUpperLimit, tcpMem[2])
+		tcpMem[2] = newUpperLimit
+		setHostTCPMemFile(hostTCPMemFile, tcpMem)
+	}
+	return nil
+}
+
+func setCg1TCPMem(podUID, containerID string, memLimit, memTCPLimit int64) error {
+	newMemTCPLimit := memLimit / 100 * int64(sockMemConfig.cgroupTCPMemRatio)
+	newMemTCPLimit = alignToPageSize(newMemTCPLimit)
+	if newMemTCPLimit < cgroupTCPMemMin2G {
```

Comment:
maybe we can also use [Clamp](https://github.com/kubewharf/katalyst-core/blob/main/pkg/util/general/common.go#L339) util function here

Reply:
Done.